### PR TITLE
Adding hasEaten to base agent struct

### LINKED ISSUE #26: Agents can keep taking food in each tick since platform stays for multiple ticks
This behaviour is maybe unrealistic - an agent can theoretically take as much food as they like but realistically they will still only be able to eat a limited amount of food before the platform moves on.

We could either limit the amount of food is taken per tick (since they can take food on multiple ticks), or introduce a `HasEaten` field which is set to true after the agent takes as much as food as they want in a single tick, and then they don't take any food on the subsequent ticks.

### PR DESCRIPTION
**Pull request by Rhys-J**
Implements only allowing agents to take food once per day. Parameter is reset in hpDecay function.

# Summary

Adding hasEaten parameter to Base Agent structure. Prevents agents from taking food multiple times per day.

Added functions to obtain or update value of hasEaten 

Alter if statement in TakeFood to include check for !hasEaten

## Additional Information

Closes #26 

## Test Plan

Added new logs for when food is taken from the platform

### CODE REVIEW

File: cmd/backend/main.go [hunk 1]
```diff
@@ -63,7 +63,7 @@ func main() {
 	foodOnPlatform := 100.0
 	numOfAgents := []int{5, 5}
 	agentHP := 100
-	days := 1
+	days := 4
```
Comment (jzzheng22) [inline cmd/backend/main.go:66]:
```suggestion
	days := 10
```

File: pkg/infra/baseagent.go [hunk 2]
```diff
@@ -85,11 +87,21 @@ func (a *Base) updateHP(foodTaken float64) {
 	a.hp = int(math.Min(100, float64(a.hp)+foodTaken))
 }
 
+func (a *Base) HasEaten() bool {
+	return a.hasEaten
+}
+
+func (a *Base) UpdateHasEaten(newStatus bool) {
```
Comment (jzzheng22) [inline pkg/infra/baseagent.go:94]:
```suggestion
func (a *Base) setHasEaten(newStatus bool) {
```

File: pkg/infra/baseagent.go [hunk 3]
```diff
@@ -85,11 +87,21 @@ func (a *Base) updateHP(foodTaken float64) {
...
+	a.hasEaten = newStatus
+}
+
 func (a *Base) TakeFood(amountOfFood float64) float64 {
-	if a.floor == a.tower.currPlatFloor {
+	if a.floor == a.tower.currPlatFloor && !a.HasEaten() {
 		foodTaken := math.Min(a.tower.currPlatFood, amountOfFood)
 		a.updateHP(foodTaken)
 		a.tower.currPlatFood -= foodTaken
+		a.UpdateHasEaten(true)
```
Comment (jzzheng22) [inline pkg/infra/baseagent.go:103]:
```suggestion
		a.setHasEaten(true)
```

File: pkg/infra/tower.go [hunk 4]
```diff
@@ -104,6 +105,7 @@ func (t *Tower) hpDecay() {
 	// TODO: can add a parameter
 	for _, agent := range t.agents {
 		newHP := agent.HP() - 20
+		agent.UpdateHasEaten(false)
```
Comment (jzzheng22) [inline pkg/infra/tower.go:108]:
```suggestion
		agent.setHasEaten(false)
```

**Review by jzzheng22 — CHANGES_REQUESTED** [review]

**Comment by jzzheng22** [comment]
I _think_ that if your function starts with a capital letter then it's exported meaning that the custom agents can call `setHasEaten()` which we don't want to do.

**Review by jzzheng22 — APPROVED** [review]

File: pkg/infra/baseagent.go [hunk 5]
```diff
@@ -85,11 +87,21 @@ func (a *Base) updateHP(foodTaken float64) {
...
+	return a.hasEaten
+}
+
+func (a *Base) setHasEaten(newStatus bool) {
+	a.hasEaten = newStatus
+}
+
 func (a *Base) TakeFood(amountOfFood float64) float64 {
-	if a.floor == a.tower.currPlatFloor {
+	if a.floor == a.tower.currPlatFloor && !a.HasEaten() {
```
Comment (JaafarRammal) [inline pkg/infra/baseagent.go:99]:
```suggestion
	if a.floor == a.tower.currPlatFloor && !a.hasEaten {
```
Custom agent can implement `HasEaten()` and not sure what the behavior would be

File: pkg/infra/tower.go [hunk 6]
```diff
@@ -48,6 +48,7 @@ func NewTower(currPlatFood float64, currPlatFloor, agentCount,
 
 func (t *Tower) Tick() {
 	//logs
+	t.Log("Reporting current platform floor", Fields{"floor": t.currPlatFloor})
```
Comment (JaafarRammal) [inline pkg/infra/tower.go:51]:
```suggestion
```

File: pkg/infra/tower.go [hunk 7]
```diff
@@ -48,6 +48,7 @@ func NewTower(currPlatFood float64, currPlatFloor, agentCount,
...
 	t.Log("Reporting food left on platform", Fields{"food_left": t.currPlatFood})
```
Comment (JaafarRammal) [inline pkg/infra/tower.go:52]:
```suggestion
	t.Log("Reporting platform status", Fields{"food_left": t.currPlatFood, "floor": t.currPlatFloor})
```

**Review by JaafarRammal — CHANGES_REQUESTED** [review]
LGTM, just minor changes

**Review by JaafarRammal — APPROVED** [review]

**Review by jzzheng22 — APPROVED** [review]
Thanks Rhys :)

**Pull request merged**